yt/data_objects/selection_objects/data_selection_objects.py

```diff
@@ -266,7 +277,7 @@ def _generate_fields(self, fields_to_generate):
                                 fi.name,
                                 sunits or "dimensionless",
                             )
-                        elif fi.dimensions != dimensions:
+                        elif not dimensions_compare_equal(fi.dimensions, dimensions):
```

Comment:
I think this is the right thing to do.  *However*, I will note that there's the possibility that this will add on a non-negligible amount of time to dataset instantiation during field detection.  I think that is unlikely, but it is a possibility.

Reply:
I'll keep it in mind.
I also want to note that field detection is time-consuming already: in tests that create a small dataset and immediately throw it away, it's adds a couple hundreds ms to each test. I think it'd be a good idea to make it lazy in the future.